### ISSUE TITLE
[Question] How to implement focal loss on ssd300 model

### ISSUE THREAD

**Issue by vkherde** [pierluigiferrari/ssd_keras#248]
Hi @pierluigiferrari 

Thanks for this exceptional tutorial, it provides a great learning. I wanted to implement Focal loss on ssd300 model which i trained on my own dataset. Can you suggest a way how can I do that. I tried implementing using below code but it's not working upto the mark:

def focal_loss(y_true, y_pred, gamma=2.0, alpha = 1.0):
     #y_pred (tensor of shape (batch size, num_boxes, num_classes)
      eps = K.epsilon()
      y_pred = K.clip(y_pred, eps, 1.-eps)
      pt = tf.where(tf.equal(y_true, 1), y_pred, 1- y_pred)
      loss = -K.pow(1.-pt, gamma) * K.log(pt)
      loss = alpha * loss
     return tf.reduce_sum(loss, axis = -1)

If you get a chance please suggest, many thanks indeed 

Regards

**Comment by DonghoonPark12** [comment]
@vkherde Hi. I also trying what you want to do and trying to understand what I need to change.
Let's try and share some intuition.

By the way, make sure that When you apply focal loss, you need to remove 'hard negative mining' parts.
Refer to it. [links](https://github.com/ailias/Focal-Loss-implement-on-Tensorflow/issues/5#issue-307638842)

**Comment by vkherde** [comment]
@DonghunP Hi, sorry for delayed response, while implementing focal loss I just changed the log loss function with above code to fit focal loss. I am not sure whether we need to remove the hard negative mining part because it is used to find appropriate bounding box for the object whereas the focal loss would only improve the classification loss for imbalanced classes. Please refer this [link](https://blog.zenggyu.com/en/post/2018-12-05/retinanet-explained-and-demystified/)

**Comment by DonghoonPark12** [comment]
@vkherde Thanks. I may misunderstand the concept.
Anyway when I implement below code(**when alpha is not 0**) loss become infinite, but when I use your code above(**when alpha is 1**) loss doesn't go infinite. Thanks anyway.

```python

def focal_loss(gamma=2, alpha=0.75):
	def focal_loss_fixed(y_true, y_pred):
		eps = K.epsilon()
		y_pred=K.clip(y_pred,eps,1.-eps)
        	pt_1 = tf.where(tf.equal(y_true, 1), y_pred, tf.ones_like(y_pred))
        	pt_0 = tf.where(tf.equal(y_true, 0), y_pred, tf.zeros_like(y_pred))
        	return -K.sum(alpha * K.pow(1. - pt_1, gamma) * K.log(pt_1))-K.sum((1-alpha) * K.pow( pt_0, gamma) * K.log(1. - pt_0))
    return focal_loss_fixed
```
